src/thermo/Phase.cpp

```diff
@@ -172,20 +173,47 @@ void Phase::getAtoms(size_t k, double* atomArray) const
     }
 }
 
+size_t Phase::findSpeciesLower(const std::string& name) const
+{
+    size_t loc = npos;
+    std::string nLower = toLowerCopy(name);
+    for (const auto& k : m_speciesIndices) {
+        if (toLowerCopy(k.first) == nLower) {
+            if (loc == npos) {
+                loc = k.second;
+            } else {
```

Comment:
Rather than falling back to an O(N) search anytime a non-canonical species name is provided, I wonder if the following might be a cleaner solution:
- Add a map of lower-case species names to species indices (say, `m_speciesIndicesLower`)
- When adding species, if there is already an entry in this map with the same name, set its value to `npos`, and possibly set the `m_caseSensitiveSpecies` flag.
- The implementation of `speciesIndex` can then use either `m_speciesIndicesLower` or `m_speciesIndices` depending on the value of `m_caseSensitiveSpecies`. If `m_speciesIndicesLower` is checked and contains `npos`, raise the exception about non-unique species identifiers.

Reply:
I am not sure about this suggestion. The second (lower-case) species map would contain nearly identical information; in cases like this, I usually keep the information in a single representation, as it is easier to maintain.

While the suggested implementation would be faster *if a user opts to use species names with names other than those specified in the input file*, I doubt that it has significant real-world benefits. Assuming that those searches are typically done during problem setup, there is little computational overhead (in repeated operations, it is _always_ faster to select species by index). I am also unsure about auto-magically flipping flags, as those decisions may be best left to the user? Weighing cost and benefit, I am suggesting to leave the implementation as is.

Reply:
True, there is a benefit to not introducing a new data structure. However, avoiding that requires several instances of this kind of convoluted try/catch construct (with I think another one needed in `Phase::species(string)`.

I think using the extra data structure would allow us to avoid adding `findSpeciesLower` as a new method -- its logic could just be in `speciesIndex`.

There's also some value in not generating exceptions for circumstances that aren't really exceptional (at least in C++). I'm less concerned about the performance implication than with the ability to run in gdb and break at an error by simply doing `catch throw`. If we use exceptions in places like this, then you end up spending a lot more time continuing past these unexceptional exceptions.

I would also note that `speciesIndex` is actually called with a name not in `m_speciesIndices` is more often than you might think -- it's not just instances of case-mismatched species names. It is used extensively in multiphase systems as a way of finding which phase a species belongs to. See for example `Kinetics::kineticsSpeciesIndex` and `Kinetics::speciesPhase`.

My thought on automatically enabling case sensitivity when there are species that can only be resolved this way was that that might be less surprising to the user than having to enable it explicitly. There may also be issues with initializing phases with species that are only distinguishable by case. For instance, what if a reaction is specified (in the input file) that contains one of these species? Adding the reaction requires looking up the indices of the involved species, and this needs to happen before the user has the chance to set the case sensitive species flag.

Reply:
@speth ... thanks for your comments (plus catching the other access-species-by-name case for `Phase::species`). I.e. this involves another (almost) duplicate map. 

Your comment regarding C++ exceptions makes a lot of sense. In the current revision, I opted to replace `std::map::at` by `std::map::find`, which eliminates the `try` - `catch` blocks. 

I do realize that the kinetics managers make extensive use of `speciesIndex` (they have to), but in most cases the logic will employ case-sensitive species without additional overhead.

Regarding the automatic flags: my thought was that in most cases, case sensitive species are present and the fallback solution is rarely used. I updated the error message to explicitly direct users towards the `caseSensitiveSpecies` flag, which can be employed after all species and reactions are set up: internally, everything is case sensitive, i.e. lowercase names are treated as fall-back solutions.

Reply:
Regarding the "automatic flags", I had to play around with this a bit to realize how this was now working - an exception is only thrown when case sensitivity is disabled and you specify a name which is not a case-sensitive match for any species. So my concern about setting up reactions is moot, and I agree that there is no need to automatically toggle this flag.

I still think the `speciesIndex` calls within the Kinetics managers for interface phases will frequently incur the O(N) overhead. For example, if you have a kinetics manager containing a surface, a gas, and a solid phase in that order, looking up the index of a species in the solid phase, even with the correct case, will cause a search through all the surface phase species, then all the gas phase species, before finally taking the easy path for the index into the solid phase.